/*************************************************************************
*                                                                        *
*   Module: OutputR                                                      *
*                                                                        *
*   Management: generate R scripts to produce the plots                  *
*                                                                        *
*   This file is part of the SeqCode 1.0 distribution                    *
*                                                                        *
*     Copyright (C) 2020 - Enrique BLANCO GARCIA                         *
*                                                                        *
*  This program is free software; you can redistribute it and/or modify  *
*  it under the terms of the GNU General Public License as published by  *
*  the Free Software Foundation; either version 2 of the License, or     *
*  (at your option) any later version.                                   *
*                                                                        *
*  This program is distributed in the hope that it will be useful,       *
*  but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*  GNU General Public License for more details.                          *
*                                                                        *
*  You should have received a copy of the GNU General Public License     *
*  along with this program; if not, write to the Free Software           * 
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.             *
*************************************************************************/

#include "seqcode/seqcode.h"

extern int VRB;
extern int WINDOWRES;
extern int SPIKEIN;
extern int COLORIZE;
extern char BGCOLOR[MAXCOLORNAME];
extern char FGCOLOR[MAXCOLORNAME];
extern char FGCOLORB[MAXCOLORNAME];
extern char GBGCOLOR[MAXCOLORNAME];
extern char GFGCOLOR[MAXCOLORNAME];
extern char HBGCOLOR[MAXCOLORNAME];
extern int UNIFORM;
extern char COLOR1[MAXCOLORNAME];
extern char COLOR2[MAXCOLORNAME];
extern char COLOR3[MAXCOLORNAME];
extern char COLOR4[MAXCOLORNAME];
extern char COLOR5[MAXCOLORNAME];
extern char COLOR6[MAXCOLORNAME];
extern char COLOR7[MAXCOLORNAME];
extern int GRADIENT;
extern int SIMPLIFIED;

extern account* m;

/****************************************************************/
/** Functions to generate the Rscripts that produce the plots  **/
/****************************************************************/

/* TSS profiles */
void CreateRscript(char* OutputFileName,
		   char* RscriptFileName,
		   char* PlotFileName,
		   char* TrackName,
		   int L,
		   long nreads,
		   long nGenes)
{
  /* Output file */
  FILE* file;

  int a,b,c;

  /* Messages */
  char mess[MAXSTRING];
    
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. R commands to produce the TSS average profile plot */
  fprintf(file,"pdf(\"%s\",bg=\"%s\",fg=\"%s\")\n",PlotFileName,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- read.table (\"%s\")\n",OutputFileName);
  
  if (UNIFORM)
    { fprintf(file,"c[,2]<-c[,2]/max(c[,2])\n"); }

  if (SPIKEIN==0)
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"TSS\",ylab=\"%s normalized count of reads\",main=\"TSS profile for %s (%d bp, %ld reads)\",pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,GFGCOLOR,GFGCOLOR,GFGCOLOR);
    }
  else
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"TSS\",ylab=\"%s normalized count of reads\",main=\"TSS profile for %s (%d bp, %ld/%d reads))\",pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,SPIKEIN,GFGCOLOR,GFGCOLOR,GFGCOLOR);      
    }
  
  a=-(L+L/5);
  b=L+(L/5);
  c=L/5;

  fprintf(file,"polygon(c(%d,%d,%d,%d),c(0,max(c[,2])+%d,max(c[,2])+%d,0),col=\"%s\")\n",
	  a,a,b,b,c,c,BGCOLOR);
  
  fprintf(file,"lines(c,col=\"%s\",lwd=8)\n",FGCOLOR);
  fprintf(file,"lines(c(0,0),c(0,max(c[,2])),lwd=1,lty=2,col=\"%s\")\n",GFGCOLOR);
  
  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"box(col=\"%s\")\n",GFGCOLOR);

  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(5.3,10,\"Figure legend. Average distribution of ChIPseq reads around the TSS of the genes.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.35,9.5,\"This plot is generated by counting the number of reads along this region for each gene and\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.4,9,\"averaging this value for the number of genes and the number of mapped reads (in millions).\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.1,8.5,\"The X-axis represents the region around the TSS in which the counts were calculated\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.3,8,\"for each gene. The Y-axis represents the intensity of the average ChIP signal normalized\",font=1,cex=0.75)\n");
  fprintf(file,"text(4.45,7.5,\"by the number of reads of the sample. TSS is the Transcription Start Site.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5.5,\"ChIPseq experiment:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5,\"%s\",font=3,cex=0.75)\n",TrackName);

  if (SPIKEIN==0)
    {
      fprintf(file,"text(5,4.5,\"Number of reads:\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld\",font=3,cex=0.75)\n",nreads);
    }
  else
    {
      fprintf(file,"text(5,4.5,\"Number of reads (spike-in):\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld (%d)\",font=3,cex=0.75)\n",nreads,SPIKEIN);      
    }
  
  fprintf(file,"text(5,3.5,\"Number of genes:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,3,\"%d\",font=3,cex=0.75)\n",m->nGenes);
  fprintf(file,"text(5,2.5,\"Flanking sequence:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,2,\"%d\",font=3,cex=0.75)\n",L);
  fprintf(file,"text(5,1.5,\"RefSeq transcripts:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,1,\"%ld\",font=3,cex=0.75)\n",m->nTotalTranscripts);
  fprintf(file,"text(5,0.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,0,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"\n");
  
  fprintf(file,"dev.off()\n\n");
  fclose(file);
}

/* Combined TSS profiles */
void CreateRscriptCombined(char* OutputFileName,
			   char* RscriptFileName,
			   char* PlotFileName,
			   char* TrackName1,
			   char* TrackName2,
			   int L,
			   long nreads1,
			   long nreads2,
			   long nGenes)
{
  /* Output file */
  FILE* file;

  int a,b,c;

  /* Messages */
  char mess[MAXSTRING];
    
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. R commands to produce the TSS average profile plot */
  fprintf(file,"pdf(\"%s\",bg=\"%s\",fg=\"%s\")\n",PlotFileName,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- read.table (\"%s\")\n",OutputFileName);
  
  if (UNIFORM)
    { fprintf(file,"c[,2]<-c[,2]/max(c[,2])\n"); }
  
  fprintf(file,"plot(c[,1],c[,2],xlab=\"TSS\",ylab=\"%s-%s normalized count of reads\",main=\"TSS profile for %s-%s (%d bp, %ld-%ld reads)\",pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName1,TrackName2,TrackName1,TrackName2,L,nreads1,nreads2,GFGCOLOR,GFGCOLOR,GFGCOLOR);
      
  a=-(L+L/5);
  b=L+(L/5);
  c=L/5;

  fprintf(file,"polygon(c(%d,%d,%d,%d),c(0,max(c[,2])+%d,max(c[,2])+%d,0),col=\"%s\")\n",
	  a,a,b,b,c,c,BGCOLOR);
  
  fprintf(file,"lines(c,col=\"%s\",lwd=8)\n",FGCOLOR);
  fprintf(file,"lines(c(0,0),c(0,max(c[,2])),lwd=1,lty=2,col=\"%s\")\n",GFGCOLOR);

  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"box(col=\"%s\")\n",GFGCOLOR);

  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(5.3,10,\"Figure legend. Combined distribution of ChIPseq reads around the TSS of the genes.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.3,9.5,\"This plot is generated by counting the number of reads along this region for each gene and\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.3,9,\"averaging this value from both experiments for the number of genes and reads (in millions).\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.2,8.5,\"The X-axis represents the region around the TSS in which the counts were calculated for\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.325,8,\"each gene. The Y-axis represents the average ChIP signal intensity subtraction normalized\",font=1,cex=0.75)\n");
  fprintf(file,"text(4.5,7.5,\"by the number of reads of each sample. TSS is the Transcription Start Site.\",font=1,cex=0.75)\n");

  fprintf(file,"text(3,5.5,\"ChIPseq experiment 1:\",font=1,cex=0.75)\n");
  fprintf(file,"text(3,5,\"%s\",font=3,cex=0.75)\n",TrackName1);
  fprintf(file,"text(3,4.5,\"Number of reads:\",font=1,cex=0.75)\n");
  fprintf(file,"text(3,4,\"%ld\",font=3,cex=0.75)\n",nreads1);

  fprintf(file,"text(7,5.5,\"ChIPseq experiment 2:\",font=1,cex=0.75)\n");
  fprintf(file,"text(7,5,\"%s\",font=3,cex=0.75)\n",TrackName2);
  fprintf(file,"text(7,4.5,\"Number of reads:\",font=1,cex=0.75)\n");
  fprintf(file,"text(7,4,\"%ld\",font=3,cex=0.75)\n",nreads2);

  fprintf(file,"text(5,3.5,\"Number of genes:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,3,\"%d\",font=3,cex=0.75)\n",m->nGenes);
  fprintf(file,"text(5,2.5,\"Flanking sequence:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,2,\"%d\",font=3,cex=0.75)\n",L);

  fprintf(file,"text(5,1.5,\"RefSeq transcripts:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,1,\"%ld\",font=3,cex=0.75)\n",m->nTotalTranscripts);
  fprintf(file,"text(5,0.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,0,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"\n");
  
  fprintf(file,"dev.off()\n\n");

  fclose(file);
}

/* GENE profiles */
void CreateRscriptGENE(char* OutputFileName,
		       char* RscriptFileName,
		       char* PlotFileName,
		       char* TrackName,
		       int L,
		       long nreads)
{
  /* Output file */
  FILE* file;

  int a,b,c;

  /* Messages */
  char mess[MAXSTRING];
    
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. Ideal gene average profile plot */
  fprintf(file,"pdf(\"%s\",bg=\"%s\",fg=\"%s\")\n",PlotFileName,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- read.table (\"%s\")\n",OutputFileName);

  if (UNIFORM)
    { fprintf(file,"c[,2]<-c[,2]/max(c[,2])\n"); }

  if (SPIKEIN==0)
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"GENE\",ylab=\"%s normalized count of reads\",main=\"GENE profile for %s (%d bp, %ld reads)\",axes=FALSE,pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,GFGCOLOR,GFGCOLOR,GFGCOLOR); 
    }
  else
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"GENE\",ylab=\"%s normalized count of reads\",main=\"GENE profile for %s (%d bp, %ld/%d reads)\",axes=FALSE,pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,SPIKEIN,GFGCOLOR,GFGCOLOR,GFGCOLOR); 
    }

  a=-(L+L/5);
  b=3*L+(L/5);
  c=L/5;

  fprintf(file,"polygon(c(%d,%d,%d,%d),c(0,max(c[,2])+%d,max(c[,2])+%d,0),col=\"%s\")\n",
	  a,a,b,b,c,c,BGCOLOR);

  fprintf(file,"lines(c,col=\"%s\",lwd=8)\n",FGCOLOR);
  fprintf(file,"lines(c(0,0),c(0,max(c[,2])),lwd=1,lty=2,col=\"%s\")\n",GFGCOLOR);
  fprintf(file,"lines(c(%d,%d),c(0,max(c[,2])),lwd=1,lty=2,col=\"%s\")\n",2*L,2*L,GFGCOLOR);

  fprintf(file,"par(font = 2,col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",
	  GFGCOLOR,GFGCOLOR,GFGCOLOR);
  fprintf(file,"axis(1,at=c(-%d,0,%d,%d),labels=c(\"-%d\",\"TSS\",\"TES\",\"+%d\"),col=\"%s\")\n",L,2*L,3*L,L,L,GFGCOLOR);
  fprintf(file,"axis(2,col=\"%s\")\n",GFGCOLOR);
  fprintf(file,"box(col=\"%s\")\n",GFGCOLOR);
  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);

  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(5.25,10,\"Figure legend. Average distribution of ChIPseq reads along a uniform gene model.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.35,9.5,\"This plot is generated by counting the number of reads along this region for each gene and\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.4,9,\"averaging this value for the number of genes and the number of mapped reads (in millions).\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.1,8.5,\"The X-axis represents the metagene and the flanking region in which the counts were\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.225,8,\"calculated for each gene. The Y-axis represents the intensity of the average ChIP signal\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.25,7.5,\"normalized by the number of reads of the sample. TSS is the Transcription Start Site and\",font=1,cex=0.75)\n");
  fprintf(file,"text(2.375,7,\"TES is the Transcription End Site.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5.5,\"ChIPseq experiment:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5,\"%s\",font=3,cex=0.75)\n",TrackName);

  if (SPIKEIN==0)
    {
      fprintf(file,"text(5,4.5,\"Number of reads:\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld\",font=3,cex=0.75)\n",nreads);
    }
  else
    {
      fprintf(file,"text(5,4.5,\"Number of reads (spike-in):\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld (%d)\",font=3,cex=0.75)\n",nreads,SPIKEIN);      
    }
    
  fprintf(file,"text(5,3.5,\"Number of genes:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,3,\"%d\",font=3,cex=0.75)\n",m->nGenes);
  fprintf(file,"text(5,2.5,\"Flanking sequence:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,2,\"%d\",font=3,cex=0.75)\n",L);
  fprintf(file,"text(5,1.5,\"RefSeq transcripts:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,1,\"%ld\",font=3,cex=0.75)\n",m->nTotalTranscripts);
  fprintf(file,"text(5,0.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,0,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  
  fprintf(file,"dev.off()\n\n");
  fclose(file);
}

/* TES profiles */
void CreateRscriptTES(char* OutputFileName,
		      char* RscriptFileName,
		      char* PlotFileName,
		      char* TrackName,
		      int L,
		      long nreads,
		      long nGenes)
{
  
  /* Output file */
  FILE* file;

  int a,b,c;

  /* Messages */
  char mess[MAXSTRING];
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. R commands to produce the TES average profile plot */
  fprintf(file,"pdf(\"%s\",bg=\"%s\",fg=\"%s\")\n",PlotFileName,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- read.table (\"%s\")\n",OutputFileName);

  if (UNIFORM)
    { fprintf(file,"c[,2]<-c[,2]/max(c[,2])\n"); }

 if (SPIKEIN==0)
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"TES\",ylab=\"%s normalized count of reads\",main=\"TES profile for %s (%d bp, %ld reads)\",pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,GFGCOLOR,GFGCOLOR,GFGCOLOR);
    }
  else
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"TES\",ylab=\"%s normalized count of reads\",main=\"TES profile for %s (%d bp, %ld/%d reads)\",pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,SPIKEIN,GFGCOLOR,GFGCOLOR,GFGCOLOR);
    }

  a=-(L+L/5);
  b=L+(L/5);
  c=L/5;

  fprintf(file,"polygon(c(%d,%d,%d,%d),c(0,max(c[,2])+%d,max(c[,2])+%d,0),col=\"%s\")\n",
	  a,a,b,b,c,c,BGCOLOR);
  
  fprintf(file,"lines(c,col=\"%s\",lwd=8)\n",FGCOLOR);
  fprintf(file,"lines(c(0,0),c(0,max(c[,2])),lwd=1,lty=2,col=\"%s\")\n",GFGCOLOR);

  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"box(col=\"%s\")\n",GFGCOLOR);

  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(5.275,10,\"Figure legend. Average distribution of ChIPseq reads around the TES of the genes.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.35,9.5,\"This plot is generated by counting the number of reads along this region for each gene and\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.4,9,\"averaging this value for the number of genes and the number of mapped reads (in millions).\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.1,8.5,\"The X-axis represents the region around the TES in which the counts were calculated\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.3,8,\"for each gene. The Y-axis represents the intensity of the average ChIP signal normalized\",font=1,cex=0.75)\n");
  fprintf(file,"text(4.425,7.5,\"by the number of reads of the sample. TES is the Transcription End Site.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5.5,\"ChIPseq experiment:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5,\"%s\",font=3,cex=0.75)\n",TrackName);

  if (SPIKEIN==0)
    {
      fprintf(file,"text(5,4.5,\"Number of reads:\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld\",font=3,cex=0.75)\n",nreads);
    }
  else
    {
      fprintf(file,"text(5,4.5,\"Number of reads (spike-in):\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld (%d)\",font=3,cex=0.75)\n",nreads,SPIKEIN);      
    }
  
  fprintf(file,"text(5,3.5,\"Number of genes:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,3,\"%d\",font=3,cex=0.75)\n",m->nGenes);
  fprintf(file,"text(5,2.5,\"Flanking sequence:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,2,\"%d\",font=3,cex=0.75)\n",L);
  fprintf(file,"text(5,1.5,\"RefSeq transcripts:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,1,\"%ld\",font=3,cex=0.75)\n",m->nTotalTranscripts);
  fprintf(file,"text(5,0.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,0,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);

  fprintf(file,"dev.off()\n\n");
  
  fclose(file);
}

/* PEAK profiles */
void CreateRscriptPeak(char* OutputFileName,
		       char* RscriptFileName,
		       char* PlotFileName,
		       char* TrackName,
		       int L,
		       long nreads,
		       long nPeaks)
{
  /* Output file */
  FILE* file;

  int a,b,c;

  /* Messages */
  char mess[MAXSTRING];
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. R commands to produce the PEAK average profile plot */
  fprintf(file,"pdf(\"%s\",bg=\"%s\",fg=\"%s\")\n",PlotFileName,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- read.table (\"%s\")\n",OutputFileName);

  if (UNIFORM)
    { fprintf(file,"c[,2]<-c[,2]/max(c[,2])\n"); }

  if (SPIKEIN==0)
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"PEAK\",ylab=\"%s normalized count of reads\",main=\"PEAK profile for %s (%d bp, %ld reads)\",pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,GFGCOLOR,GFGCOLOR,GFGCOLOR);
    }
  else
    {
      fprintf(file,"plot(c[,1],c[,2],xlab=\"PEAK\",ylab=\"%s normalized count of reads\",main=\"PEAK profile for %s (%d bp, %ld/%d reads)\",pch=\" \",col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",TrackName,TrackName,L,nreads,SPIKEIN,GFGCOLOR,GFGCOLOR,GFGCOLOR);
    }
  
  a=-(L+L/5);
  b=L+(L/5);
  c=L/5;

  fprintf(file,"polygon(c(%d,%d,%d,%d),c(0,max(c[,2])+%d,max(c[,2])+%d,0),col=\"%s\")\n",
	  a,a,b,b,c,c,BGCOLOR);

  fprintf(file,"lines(c,col=\"%s\",lwd=8)\n",FGCOLOR);
  fprintf(file,"lines(c(0,0),c(0,max(c[,2])),lwd=1,lty=2,col=\"%s\")\n",GFGCOLOR);

  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"box(col=\"%s\")\n",GFGCOLOR);

  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(5.4,10,\"Figure legend. Average distribution of ChIPseq reads around the center of the peaks.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.35,9.5,\"This plot is generated by counting the number of reads along this region for each peak and\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.4,9,\"averaging this value for the number of peaks and the number of mapped reads (in millions).\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.5,8.5,\"The X-axis represents the region around the peak center in which the counts were calculated\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.3,8,\"for each peak. The Y-axis represents the intensity of the average ChIP signal normalized\",font=1,cex=0.75)\n");
  fprintf(file,"text(4.85,7.5,\"by the number of reads of the sample. PEAK represents the center of the peaks.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5.5,\"ChIPseq experiment:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,5,\"%s\",font=3,cex=0.75)\n",TrackName);

  if (SPIKEIN==0)
    {
      fprintf(file,"text(5,4.5,\"Number of reads:\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld\",font=3,cex=0.75)\n",nreads);
    }
  else
    {
      fprintf(file,"text(5,4.5,\"Number of reads (spike-in):\",font=1,cex=0.75)\n");
      fprintf(file,"text(5,4,\"%ld (%d)\",font=3,cex=0.75)\n",nreads,SPIKEIN);      
    }
  
  fprintf(file,"text(5,3.5,\"Number of peaks:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,3,\"%ld\",font=3,cex=0.75)\n",m->nPeaks);
  fprintf(file,"text(5,2.5,\"Flanking sequence:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,2,\"%d\",font=3,cex=0.75)\n",L);
  fprintf(file,"text(5,1.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(5,1,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"dev.off()\n\n");

  fclose(file);
}

/* TSS heat maps */
void CreateRscriptMaps(char* OutputFileName2,
		       char* RscriptFileName,
		       char* PlotFileName2,
		       char* TrackName,
		       int L,
		       long nreads,
		       long nGenes)
{
  /* Output file */
  FILE* file;

  /* Messages */
  char mess[MAXSTRING];
    
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. R commands to produce the TSS map plot */
  fprintf(file,"pdf(\"%s\",width=5,height=10,bg=\"%s\",fg=\"%s\")\n",
	  PlotFileName2,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- scan(\"%s\",sep=\"\\t\")\n",OutputFileName2);

  /* Increase the intensity of colors */
  if (COLORIZE)
    {
      fprintf(file,"c[c>0]=10\n");
    }
  
  fprintf(file,"m<-matrix(c,%ld,%d,byrow=TRUE)\n",nGenes,(2*L)/WINDOWRES);
  fprintf(file,"min <- min(c, na.rm=T)\n");
  fprintf(file,"max <- max(c, na.rm=T)\n");
  fprintf(file,"code1 <- col2rgb(\"%s\")\n",BGCOLOR);
  fprintf(file,"code2 <- col2rgb(\"%s\")\n",FGCOLOR);
  
  fprintf(file,"ColorRamp <- rgb(seq(code1[1]/255,code2[1]/255,length=50),  # Red\nseq(code1[2]/255,code2[2]/255,length=50),  # Green\nseq(code1[3]/255,code2[3]/255,length=50))  # Blue\n");

  fprintf(file,"ColorRamp[1]=\"%s\"\n",HBGCOLOR);

  fprintf(file,"ColorLevels <- seq(min/max, max/max, length=length(ColorRamp))\n");

  fprintf(file,"layout(matrix(data=c(1,2), nrow=1, ncol=2), widths=c(4,1),heights=c(1,1))\n");
  fprintf(file,"par(mar = c(5,5,2.5,1), font = 2,col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",GFGCOLOR,GFGCOLOR,GFGCOLOR);
  
  fprintf(file,"image(1:ncol(m), 1:nrow(m), t(m),col=ColorRamp, xlab=\"REGION\", ylab=\"GENES\",\n");
  fprintf(file,"axes=FALSE, zlim=c(min,max), main= \"%s\")\n",TrackName);
  
  fprintf(file,"axis(1,at=c(1,%f,%d),labels=c(\"-%d\",\"TSS\",\"+%d\"))\n",
	  (L/WINDOWRES)+0.5,
	  2*L/WINDOWRES,
	  L,L);
  fprintf(file,"axis(2,at=c(1,%ld),labels=c(\"1\",\"%ld\"))\n",nGenes,nGenes);
  
  fprintf(file,"par(mar = c(5,2.5,2.5,1))\n");
  fprintf(file,"image(1, ColorLevels, matrix(data=ColorLevels, ncol=length(ColorLevels),nrow=1),\n");
  fprintf(file,"col=ColorRamp,xlab=\"\",ylab=\"\",xaxt=\"n\", las = 1)\n");
  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);

  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(1,5,\"Figure legend. Heatmap of ChIPseq reads around the TSS of the genes.\",font=2,cex=0.75,srt=90)\n");
  fprintf(file,"text(1.5,4.95,\"This plot is generated by counting the number of reads along this region for\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2,4.85,\"each gene to display these values on different rows of the heatmap. The\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2.5,4.825,\"X-axis represents the region around the TSS in which the counts were\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3,4.95,\"calculated for each gene. The Y-axis represents the list of genes ranked by\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3.5,5,\"the average ChIP intensity on this region. TSS is the Transcription Start Site.\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(4,4.5,\"On the right, ChIP signal range from high (top) to low (bottom).\",font=1,cex=0.75,srt=90)\n");

  fprintf(file,"text(6,2.5,\"ChIPseq experiment:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(6.5,2.5,\"%s\",font=3,cex=0.75,srt=90)\n",TrackName);

 if (SPIKEIN==0)
    {
      fprintf(file,"text(6,7.5,\"Number of reads:\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld\",font=3,cex=0.75,srt=90)\n",nreads);
    }
  else
    {
      fprintf(file,"text(6,7.5,\"Number of reads (spike-in):\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld (%d)\",font=3,cex=0.75,srt=90)\n",nreads,SPIKEIN);
    }

  fprintf(file,"text(7,2.5,\"Number of genes:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,2.5,\"%d\",font=3,cex=0.75,srt=90)\n",m->nGenes);
  fprintf(file,"text(7,7.5,\"Flanking sequence:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",L);

  fprintf(file,"text(8,2.5,\"RefSeq transcripts:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,2.5,\"%ld\",font=3,cex=0.75,srt=90)\n",m->nTotalTranscripts);
  fprintf(file,"text(8,7.5,\"Window factor:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",WINDOWRES);
  fprintf(file,"text(11,5,cex=0.5,font=2,\"SeqCode by E. Blanco %s\",srt=90)\n",mess);
  
  fprintf(file,"dev.off()\n");
  
  fclose(file);
}

/* Combined TSS heat maps */
void CreateRscriptCombinedMaps(char* OutputFileName2,
			       char* RscriptFileName,
			       char* PlotFileName2,
			       char* TrackName1,
			       char* TrackName2,
			       int L,
			       long nreads1,
			       long nreads2,
			       long nGenes)
{
  /* Output file */
  FILE* file;
  
  /* Messages */
  char mess[MAXSTRING];
  
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }
  
  /* 1. R commands to produce the TSS map plot */
  fprintf(file,"pdf(\"%s\",width=5,height=10,bg=\"%s\",fg=\"%s\")\n",
	  PlotFileName2,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- scan(\"%s\",sep=\"\\t\")\n",OutputFileName2);

  /* Increase the intensity of colors */
  if (COLORIZE)
    {
      fprintf(file,"c[c>0]=10\n");
    }
  
  fprintf(file,"m<-matrix(c,%ld,%d,byrow=TRUE)\n",nGenes,(2*L)/WINDOWRES);
  fprintf(file,"min <- min(c, na.rm=T)\n");
  fprintf(file,"max <- max(c, na.rm=T)\n");

  fprintf(file,"code1 <- col2rgb(\"%s\")\n",BGCOLOR);
  fprintf(file,"code2 <- col2rgb(\"%s\")\n",FGCOLOR);
  fprintf(file,"ColorRampA <- rgb(seq(code1[1]/255,code2[1]/255,length=50),  # Red\nseq(code1[2]/255,code2[2]/255,length=50),  # Green\nseq(code1[3]/255,code2[3]/255,length=50))  # Blue\n");
  fprintf(file,"ColorRampA[1]=\"%s\"\n",HBGCOLOR);

  fprintf(file,"code3 <- col2rgb(\"%s\")\n",FGCOLORB);
  fprintf(file,"code4 <- col2rgb(\"%s\")\n",BGCOLOR);
  fprintf(file,"ColorRampB <- rgb(seq(code3[1]/255,code4[1]/255,length=50),  # Red\nseq(code3[2]/255,code4[2]/255,length=50),  # Green\nseq(code3[3]/255,code4[3]/255,length=50))  # Blue\n");
  fprintf(file,"ColorRampB[50]=\"%s\"\n",HBGCOLOR);

  fprintf(file,"ColorRamp <-c (ColorRampB,ColorRampA)\n");
  
  fprintf(file,"ColorLevels <- seq(min/max, max/max, length=length(ColorRamp))\n");
  
  fprintf(file,"layout(matrix(data=c(1,2), nrow=1, ncol=2), widths=c(4,1),heights=c(1,1))\n");
  fprintf(file,"par(mar = c(5,5,2.5,1), font = 2,col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",GFGCOLOR,GFGCOLOR,GFGCOLOR);
  
  fprintf(file,"image(1:ncol(m), 1:nrow(m), t(m),col=ColorRamp, xlab=\"REGION\", ylab=\"GENES\",\n");
  fprintf(file,"axes=FALSE, zlim=c(min,max), main= \"%s-%s\")\n",TrackName1,TrackName2);
  
  fprintf(file,"axis(1,at=c(1,%f,%d),labels=c(\"-%d\",\"TSS\",\"+%d\"))\n",
	  (L/WINDOWRES)+0.5,
	  2*L/WINDOWRES,
	  L,L);
  fprintf(file,"axis(2,at=c(1,%ld),labels=c(\"1\",\"%ld\"))\n",nGenes,nGenes);
  
  fprintf(file,"par(mar = c(5,2.5,2.5,1))\n");
  fprintf(file,"image(1, ColorLevels, matrix(data=ColorLevels, ncol=length(ColorLevels),nrow=1),\n");
  fprintf(file,"col=ColorRamp,xlab=\"\",ylab=\"\",xaxt=\"n\", las = 1)\n");
  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  
  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(1,5,\"Figure legend. Combined heat map of ChIPseq reads around the TSS of the genes.\",font=2,cex=0.75,srt=90)\n");
  fprintf(file,"text(1.5,4.95,\"This plot is generated by counting the number of reads along this region for\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2,4.85,\"each gene in both experiments to display these values on different rows of the map. The\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2.5,4.825,\"X-axis represents the region around the TSS in which the counts were\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3,4.95,\"calculated for each gene. The Y-axis represents the list of genes ranked by\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3.5,5,\"the average ChIP intensity subtraction on this region. TSS is the Transcription Start Site.\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(4,4.5,\"On the right, ChIP signal difference range from high (top) to low (bottom).\",font=1,cex=0.75,srt=90)\n");
  
  fprintf(file,"text(6,2.5,\"ChIPseq experiment 1:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(6.5,2.5,\"%s\",font=3,cex=0.75,srt=90)\n",TrackName1);
  fprintf(file,"text(6,7.5,\"Number of reads:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(6.5,7.5,\"%ld\",font=3,cex=0.75,srt=90)\n",nreads1);

  fprintf(file,"text(7,2.5,\"ChIPseq experiment 2:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,2.5,\"%s\",font=3,cex=0.75,srt=90)\n",TrackName2);
  fprintf(file,"text(7,7.5,\"Number of reads:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,7.5,\"%ld\",font=3,cex=0.75,srt=90)\n",nreads2);

  fprintf(file,"text(8,2.5,\"Number of genes:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,2.5,\"%d\",font=3,cex=0.75,srt=90)\n",m->nGenes);
  fprintf(file,"text(8,7.5,\"Flanking sequence:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",L);
  
  fprintf(file,"text(9,2.5,\"RefSeq transcripts:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(9.5,2.5,\"%ld\",font=3,cex=0.75,srt=90)\n",m->nTotalTranscripts);
  fprintf(file,"text(9,7.5,\"Window factor:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(9.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",WINDOWRES);
  fprintf(file,"text(11,5,cex=0.5,font=2,\"SeqCode by E. Blanco %s\",srt=90)\n",mess);
  
  fprintf(file,"dev.off()\n");
  
  fclose(file);
}

/* TES heat maps */
void CreateRscriptTESMaps(char* OutputFileName2,
			  char* RscriptFileName,
			  char* PlotFileName2,
			  char* TrackName,
			  int L,
			  long nreads,
			  long nGenes)
{
  /* Output file */
  FILE* file;

  /* Messages */
  char mess[MAXSTRING];
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 2. R commands to produce the TES map */
  fprintf(file,"pdf(\"%s\",width=5,height=10,bg=\"%s\",fg=\"%s\")\n",
	  PlotFileName2,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- scan(\"%s\",sep=\"\\t\")\n",OutputFileName2);

  /* Increase the intensity of colors */
  if (COLORIZE)
    {
      fprintf(file,"c[c>0]=10\n");
    }
  
  fprintf(file,"m<-matrix(c,%ld,%d,byrow=TRUE)\n",nGenes,(2*L)/WINDOWRES);
  fprintf(file,"min <- min(c, na.rm=T)\n");
  fprintf(file,"max <- max(c, na.rm=T)\n");
  fprintf(file,"code1 <- col2rgb(\"%s\")\n",BGCOLOR);
  fprintf(file,"code2 <- col2rgb(\"%s\")\n",FGCOLOR);
  fprintf(file,"ColorRamp <- rgb(seq(code1[1]/255,code2[1]/255,length=50),  # Red\nseq(code1[2]/255,code2[2]/255,length=50),  # Green\nseq(code1[3]/255,code2[3]/255,length=50))  # Blue\n");

  fprintf(file,"ColorRamp[1]=\"%s\"\n",HBGCOLOR);

  fprintf(file,"ColorLevels <- seq(min/max, max/max, length=length(ColorRamp))\n");
      
  fprintf(file,"layout(matrix(data=c(1,2), nrow=1, ncol=2), widths=c(4,1),heights=c(1,1))\n");
  fprintf(file,"par(mar = c(5,5,2.5,1), font = 2,col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",GFGCOLOR,GFGCOLOR,GFGCOLOR);
  
  fprintf(file,"image(1:ncol(m), 1:nrow(m), t(m),col=ColorRamp, xlab=\"REGION\", ylab=\"GENES\",\n");
  fprintf(file,"axes=FALSE, zlim=c(min,max), main= \"%s\")\n",TrackName);
  
  fprintf(file,"axis(1,at=c(1,%f,%d),labels=c(\"-%d\",\"TES\",\"+%d\"))\n",
	  (L/WINDOWRES)+0.5,
	  2*L/WINDOWRES,
	  L,L);
  fprintf(file,"axis(2,at=c(1,%ld),labels=c(\"1\",\"%ld\"))\n",nGenes,nGenes);
  
  fprintf(file,"par(mar = c(5,2.5,2.5,1))\n");
  fprintf(file,"image(1, ColorLevels, matrix(data=ColorLevels, ncol=length(ColorLevels),nrow=1),\n");
  fprintf(file,"col=ColorRamp,xlab=\"\",ylab=\"\",xaxt=\"n\", las = 1)\n");
  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);

  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(1,5,\"Figure legend. Heatmap of ChIPseq reads around the TES of the genes.\",font=2,cex=0.75,srt=90)\n");
  fprintf(file,"text(1.5,4.9275,\"This plot is generated by counting the number of reads along this region for\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2,4.85,\"each gene to display these values on different rows of the heatmap. The\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2.5,4.8,\"X-axis represents the region around the TES in which the counts were\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3,4.95,\"calculated for each gene. The Y-axis represents the list of genes ranked by\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3.5,4.95,\"the average ChIP intensity on this region. TES is the Transcription End Site.\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(4,4.5,\"On the right, ChIP signal range from high (top) to low (bottom).\",font=1,cex=0.75,srt=90)\n");

  fprintf(file,"text(6,2.5,\"ChIPseq experiment:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(6.5,2.5,\"%s\",font=3,cex=0.75,srt=90)\n",TrackName);

  if (SPIKEIN==0)
    {
      fprintf(file,"text(6,7.5,\"Number of reads:\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld\",font=3,cex=0.75,srt=90)\n",nreads);
    }
  else
    {
      fprintf(file,"text(6,7.5,\"Number of reads (spike-in):\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld (%d)\",font=3,cex=0.75,srt=90)\n",nreads,SPIKEIN);
    }

  fprintf(file,"text(7,2.5,\"Number of genes:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,2.5,\"%d\",font=3,cex=0.75,srt=90)\n",m->nGenes);
  fprintf(file,"text(7,7.5,\"Flanking sequence:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",L);

  fprintf(file,"text(8,2.5,\"RefSeq transcripts:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,2.5,\"%ld\",font=3,cex=0.75,srt=90)\n",m->nTotalTranscripts);
  fprintf(file,"text(8,7.5,\"Window factor:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",WINDOWRES);
  fprintf(file,"text(11,5,cex=0.5,font=2,\"SeqCode by E. Blanco %s\",srt=90)\n",mess);
  
  fprintf(file,"dev.off()\n");
  
  fclose(file);
}

/* GENE heat maps */
void CreateRscriptGENEMaps(char* OutputFileName2,
			   char* RscriptFileName,
			   char* PlotFileName2,
			   char* TrackName,
			   int L,
			   long nreads,
			   long nGenes)
{
  
  /* Output file */
  FILE* file;

  /* Messages */
  char mess[MAXSTRING];
    
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. R commands to produce the GENE map plot */
  fprintf(file,"pdf(\"%s\",width=5,height=10,bg=\"%s\",fg=\"%s\")\n",
	  PlotFileName2,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- scan(\"%s\",sep=\"\\t\")\n",OutputFileName2);

  /* Increase the intensity of colors */
  if (COLORIZE)
    {
      fprintf(file,"c[c>0]=10\n");
    }
  
  fprintf(file,"m<-matrix(c,%ld,%d,byrow=TRUE)\n",nGenes,(4*L)/WINDOWRES);
  fprintf(file,"min <- min(c, na.rm=T)\n");
  fprintf(file,"max <- max(c, na.rm=T)\n");
  fprintf(file,"code1 <- col2rgb(\"%s\")\n",BGCOLOR);
  fprintf(file,"code2 <- col2rgb(\"%s\")\n",FGCOLOR);
  
  fprintf(file,"ColorRamp <- rgb(seq(code1[1]/255,code2[1]/255,length=50),  # Red\nseq(code1[2]/255,code2[2]/255,length=50),  # Green\nseq(code1[3]/255,code2[3]/255,length=50))  # Blue\n");

  fprintf(file,"ColorRamp[1]=\"%s\"\n",HBGCOLOR);

  fprintf(file,"ColorLevels <- seq(min/max, max/max, length=length(ColorRamp))\n");

  fprintf(file,"layout(matrix(data=c(1,2), nrow=1, ncol=2), widths=c(4,1),heights=c(1,1))\n");
  fprintf(file,"par(mar = c(5,5,2.5,1), font = 2,col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",GFGCOLOR,GFGCOLOR,GFGCOLOR);
  
  fprintf(file,"image(1:ncol(m), 1:nrow(m), t(m),col=ColorRamp, xlab=\"REGION\", ylab=\"GENES\",\n");
  fprintf(file,"axes=FALSE, zlim=c(min,max), main= \"%s\")\n",TrackName);
  
  fprintf(file,"axis(1,at=c(1,%d,%d,%d),labels=c(\"-%d\",\"TSS\",\"TES\",\"+%d\"))\n",
	  (L/WINDOWRES)+1,
	  (3*L/WINDOWRES),
	  (4*L/WINDOWRES),
	  L,L);
  fprintf(file,"axis(2,at=c(1,%ld),labels=c(\"1\",\"%ld\"))\n",nGenes,nGenes);
  
  fprintf(file,"par(mar = c(5,2.5,2.5,1))\n");
  fprintf(file,"image(1, ColorLevels, matrix(data=ColorLevels, ncol=length(ColorLevels),nrow=1),\n");
  fprintf(file,"col=ColorRamp,xlab=\"\",ylab=\"\",xaxt=\"n\", las = 1)\n");
  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);

  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(1,5,\"Figure legend. Heatmap of ChIPseq reads along a uniform gene model.\",font=2,cex=0.75,srt=90)\n");
  fprintf(file,"text(1.5,4.975,\"This plot is generated by counting the number of reads along this region for\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2,4.975,\"each gene to display these values on different rows of the heatmap. The X-\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2.5,4.95,\"axis represents the metagene and the flanking regions in which the counts\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3,4.9,\"were calculated for each gene. The Y-axis represents the list of all genes\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3.5,5,\"ranked by the average ChIP intensity on this region. TSS is the Transcription\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(4,4.9,\"Start Site and TES is the Transcription End Site. On the right, ChIP signal\",font=1,cex=0.75,srt=90)\n");
    fprintf(file,"text(4.5,3.65,\"range from high (top) to low (bottom).\",font=1,cex=0.75,srt=90)\n");

  fprintf(file,"text(6,2.5,\"ChIPseq experiment:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(6.5,2.5,\"%s\",font=3,cex=0.75,srt=90)\n",TrackName);

  if (SPIKEIN==0)
    {
      fprintf(file,"text(6,7.5,\"Number of reads:\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld\",font=3,cex=0.75,srt=90)\n",nreads);
    }
  else
    {
      fprintf(file,"text(6,7.5,\"Number of reads (spike-in):\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld (%d)\",font=3,cex=0.75,srt=90)\n",nreads,SPIKEIN);
    }
  
  fprintf(file,"text(7,2.5,\"Number of genes:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,2.5,\"%d\",font=3,cex=0.75,srt=90)\n",m->nGenes);
  fprintf(file,"text(7,7.5,\"Flanking sequence:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",L);

  fprintf(file,"text(8,2.5,\"RefSeq transcripts:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,2.5,\"%ld\",font=3,cex=0.75,srt=90)\n",m->nTotalTranscripts);
  fprintf(file,"text(8,7.5,\"Window factor:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",WINDOWRES);
  fprintf(file,"text(11,5,cex=0.5,font=2,\"SeqCode by E. Blanco %s\",srt=90)\n",mess);
  
  fprintf(file,"dev.off()\n");
  
  fclose(file);
}

/* PEAK heat maps */
void CreateRscriptPeakMaps(char* OutputFileName4,
			   char* RscriptFileName,
			   char* PlotFileName2,
			   char* TrackName,
			   int L,
			   long nreads,
			   long nPeaks)
{
  /* Output file */
  FILE* file;

  /* Messages */
  char mess[MAXSTRING];
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* 1. R commands to generate the PEAK map */
  fprintf(file,"pdf(\"%s\",width=5,height=10,bg=\"%s\",fg=\"%s\")\n",
	  PlotFileName2,GBGCOLOR,GFGCOLOR);
  fprintf(file,"c <- scan(\"%s\",sep=\"\\t\")\n",OutputFileName4);

  /* Increase the intensity of colors */
  if (COLORIZE)
    {
      fprintf(file,"c[c>0]=10\n");
    }
  
  fprintf(file,"m<-matrix(c,%ld,%d,byrow=TRUE)\n",nPeaks,(2*L)/WINDOWRES);
  fprintf(file,"min <- min(c, na.rm=T)\n");
  fprintf(file,"max <- max(c, na.rm=T)\n");
  fprintf(file,"code1 <- col2rgb(\"%s\")\n",BGCOLOR);
  fprintf(file,"code2 <- col2rgb(\"%s\")\n",FGCOLOR);
  
  fprintf(file,"ColorRamp <- rgb(seq(code1[1]/255,code2[1]/255,length=50),  # Red\nseq(code1[2]/255,code2[2]/255,length=50),  # Green\nseq(code1[3]/255,code2[3]/255,length=50))  # Blue\n");

  fprintf(file,"ColorRamp[1]=\"%s\"\n",HBGCOLOR);

  fprintf(file,"ColorLevels <- seq(min/max, max/max, length=length(ColorRamp))\n");
      
  fprintf(file,"layout(matrix(data=c(1,2), nrow=1, ncol=2), widths=c(4,1),heights=c(1,1))\n");
  fprintf(file,"par(mar = c(5,5,2.5,1), font = 2,col.axis=\"%s\",col.main=\"%s\",col.lab=\"%s\")\n",GFGCOLOR,GFGCOLOR,GFGCOLOR);
  
  fprintf(file,"image(1:ncol(m), 1:nrow(m), t(m),col=ColorRamp, xlab=\"REGION\", ylab=\"PEAKS\",\n");
  fprintf(file,"axes=FALSE, zlim=c(min,max), main= \"%s\")\n",TrackName);
  
  fprintf(file,"axis(1,at=c(1,%d,%d),labels=c(\"-%d\",\"PEAK\",\"+%d\"))\n",L/WINDOWRES,2*L/WINDOWRES,L,L);
  fprintf(file,"axis(2,at=c(1,%ld),labels=c(\"1\",\"%ld\"))\n",nPeaks,nPeaks);
  
  fprintf(file,"par(mar = c(5,2.5,2.5,1))\n");
  fprintf(file,"image(1, ColorLevels, matrix(data=ColorLevels, ncol=length(ColorLevels),nrow=1),\n");
  fprintf(file,"col=ColorRamp,xlab=\"\",ylab=\"\",xaxt=\"n\", las = 1)\n");
  GetTime(mess,m);
  fprintf(file,"mtext(side=4,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);

  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(1,5,\"Figure legend. Heatmap of ChIPseq reads around the center of the peaks.\",font=2,cex=0.75,srt=90)\n");
  fprintf(file,"text(1.5,5.05,\"This plot is generated by counting the number of reads along this region for each\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2,4.825,\"peak to display these values on different rows of the heatmap. The X-axis\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(2.5,5.1,\"represents the region around the peak center in which the counts were calculated\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3,4.95,\"for each peak. The Y-axis represents the list of peaks ranked by the average\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(3.5,5,\"ChIP intensity on this region. PEAK represents the center of the peaks. On the\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(4,4.175,\"right, ChIP signal range from high (top) to low (bottom).\",font=1,cex=0.75,srt=90)\n");

  fprintf(file,"text(6,2.5,\"ChIPseq experiment:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(6.5,2.5,\"%s\",font=3,cex=0.75,srt=90)\n",TrackName);

  if (SPIKEIN==0)
    {
      fprintf(file,"text(6,7.5,\"Number of reads:\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld\",font=3,cex=0.75,srt=90)\n",nreads);
    }
  else
    {
      fprintf(file,"text(6,7.5,\"Number of reads (spike-in):\",font=1,cex=0.75,srt=90)\n");
      fprintf(file,"text(6.5,7.5,\"%ld (%d)\",font=3,cex=0.75,srt=90)\n",nreads,SPIKEIN);
    }

  fprintf(file,"text(7,2.5,\"Number of peaks:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,2.5,\"%ld\",font=3,cex=0.75,srt=90)\n",m->nPeaks);
  fprintf(file,"text(7,7.5,\"Flanking sequence:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(7.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",L);

  fprintf(file,"text(8,7.5,\"Window factor:\",font=1,cex=0.75,srt=90)\n");
  fprintf(file,"text(8.5,7.5,\"%d\",font=3,cex=0.75,srt=90)\n",WINDOWRES);
  fprintf(file,"text(11,5,cex=0.5,font=2,\"SeqCode by E. Blanco %s\",srt=90)\n",mess);

  fprintf(file,"dev.off()\n");

  fclose(file);
}

/* Pie charts: peak genome distribution */
void CreateRscriptChart(char* RscriptFileName,
			char* PlotFileName,
			int nPromoters,
			int nIntragenics,
			int nIntergenics,
			char* TrackName,
			long nPeaks)
{
  
  /* Output file */
  FILE* file;

  /* Total peaks */
  int nTotalPeaks;

  /* Messages */
  char mess[MAXSTRING];
    
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* Calculating the total number of peaks */
  nTotalPeaks =  nPromoters + nIntragenics + nIntergenics;

  /* Pie chart representation */
  fprintf(file,"pdf(\"%s\")\n",PlotFileName); 
  fprintf(file,"slices<-c(%d,%d,%d)\n",nPromoters,nIntragenics,nIntergenics);
  fprintf(file,"percent<-round(slices/sum(slices)*100,digits=1)\n");
  fprintf(file,"### FRAMEWORK\n");
  fprintf(file,"mat <- matrix(c(1,2),1)\n");
  fprintf(file,"layout(mat,c(2,1),c(1,1))\n");

  fprintf(file,"### LEFT PANEL\n");
  fprintf(file,"par(mar=c(0.5,0.5,0.5,0.5),lwd = 4)\n");

  if (GRADIENT)
    {
      fprintf(file,"colfunc<-colorRampPalette(c(\"%s\",\"white\"))\n",COLOR1);
      fprintf(file,"colors<-colfunc(3)\n");
      fprintf(file,"pie(slices,labels=NA, col=colors,");
    }
  else
    {
      fprintf(file,"pie(slices,labels=NA,col=c(\"%s\",\"%s\",\"%s\"),",
	      COLOR1,COLOR2,COLOR3);
    }
  fprintf(file,"init.angle=90,clockwise=TRUE,radius=1)\n");
  
  fprintf(file,"text(x=0,y=1.25,\"%s (%d peaks)\",cex=1.75,font=2)\n",TrackName,nTotalPeaks);
  GetTime(mess,m);
  fprintf(file,"text(x=0,y=-1.25,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  
  fprintf(file,"### RIGHT PANEL\n");
  fprintf(file,"plot(0,0,pch=\" \",ylim=c(0,2),xlim=c(0,5),axes=FALSE)\n");
  fprintf(file,"par(mar=c(0.5,0.5,0.5,0.5),lwd = 2)\n");

  if (GRADIENT)
    {
      fprintf(file,"lines(c(0,0.5),c(1.3,1.3),col=colors[1],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(1.1,1.1),col=colors[2],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(0.9,0.9),col=\"gray97\",lwd=8)\n");
    }
  else
    {
      fprintf(file,"lines(c(0,0.5),c(1.3,1.3),col=\"%s\",lwd=8)\n",COLOR1);
      fprintf(file,"lines(c(0,0.5),c(1.1,1.1),col=\"%s\",lwd=8)\n",COLOR2);
      fprintf(file,"lines(c(0,0.5),c(0.9,0.9),col=\"%s\",lwd=8)\n",COLOR3);
    }
  
  fprintf(file,"text(x=2.5,y=1.5,\"REGIONS\",cex=2,font=2)\n");
  fprintf(file,"text(x=2.65,y=1.3,\"PROMOTER\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[1],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=1.2,string,cex=1.5)\n");
  fprintf(file,"text(x=2.75,y=1.1,\"INTRAGENIC\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[2],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=1.0,string,cex=1.5)\n");
  fprintf(file,"text(x=2.75,y=0.9,\"INTERGENIC\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[3],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=0.8,string,cex=1.5)\n");
  fprintf(file,"rect(-0.15,0.7,5,1.6)\n");

  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(6,9,\"Figure legend. Distribution of ChIPseq reads across different regions of the genome.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.75,8.5,\"This plot is generated by counting the number of peaks fitting on each class of region.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.65,8,\"PROMOTER region is the region within 2.5 Kbp upstream of the TSS and the TSS.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.6,7.5,\"INTRAGENIC regions is the region within the transcripts. INTERGENIC is the rest\",font=1,cex=0.75)\n");
  fprintf(file,"text(4.225,7,\"of the genome. TSS is the Transcription Start Site.\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,5.5,\"ChIPseq experiment:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,5,\"%s\",font=3,cex=0.75)\n",TrackName);
  fprintf(file,"text(6,4.5,\"Number of peaks:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,4,\"%ld\",font=3,cex=0.75)\n",nPeaks);
  fprintf(file,"text(6,3.5,\"RefSeq transcripts:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,3,\"%ld\",font=3,cex=0.75)\n",m->nTotalTranscripts);
  fprintf(file,"text(6,2.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,2,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"text(6,0,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"dev.off()\n\n");

  fclose(file);
}

/* Pie charts: peak genome distribution */
void CreateRscriptChartSimplified(char* RscriptFileName,
				  char* PlotFileName,
				  int nPromoters,
				  int nIntragenics,
				  int nIntergenics,
				  char* TrackName,
				  long nPeaks)
{
  
  /* Output file */
  FILE* file;

  /* Total peaks */
  int nTotalPeaks;

  /* Messages */
  char mess[MAXSTRING];
    
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* Calculating the total number of peaks */
  nTotalPeaks =  nPromoters + nIntragenics + nIntergenics;

  /* Pie chart representation */
  fprintf(file,"pdf(\"%s\")\n",PlotFileName); 
  fprintf(file,"slices<-c(%d,%d,%d)\n",nPromoters,nIntragenics,nIntergenics);
  fprintf(file,"percent<-round(slices/sum(slices)*100,digits=1)\n");
  fprintf(file,"### FRAMEWORK\n");
  fprintf(file,"mat <- matrix(c(1,2),1)\n");
  fprintf(file,"layout(mat,c(2,1),c(1,1))\n");

  fprintf(file,"### LEFT PANEL\n");
  fprintf(file,"par(mar=c(0.5,0.5,0.5,0.5),lwd = 4)\n");

  if (GRADIENT)
    {
      fprintf(file,"colfunc<-colorRampPalette(c(\"%s\",\"white\"))\n",COLOR1);
      fprintf(file,"colors<-colfunc(3)\n");
      fprintf(file,"pie(slices,labels=NA, col=colors,");
    }
  else
    {
      /* Default blue/grey/white color scheme */
      if (!strcmp(COLOR1,"coral") && !strcmp(COLOR2,"darkred") && !strcmp(COLOR3,"burlywood"))
	{
	  strcpy(COLOR1,"blue");
	  strcpy(COLOR2,"grey");
	  strcpy(COLOR3,"white");
	}
      
      fprintf(file,"pie(slices,labels=NA,col=c(\"%s\",\"%s\",\"%s\"),",
	      COLOR1,COLOR2,COLOR3);
    }
  fprintf(file,"init.angle=90,clockwise=TRUE,radius=1)\n");
  
  fprintf(file,"text(x=0,y=1.25,\"%s (%d peaks)\",cex=1.75,font=2)\n",TrackName,nTotalPeaks);
  GetTime(mess,m);
  fprintf(file,"text(x=0,y=-1.25,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  
  fprintf(file,"### RIGHT PANEL\n");
  fprintf(file,"plot(0,0,pch=\" \",ylim=c(0,2),xlim=c(0,5),axes=FALSE)\n");
  fprintf(file,"par(mar=c(0.5,0.5,0.5,0.5),lwd = 2)\n");

  if (GRADIENT)
    {
      fprintf(file,"lines(c(0,0.5),c(1.3,1.3),col=colors[1],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(1.1,1.1),col=colors[2],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(0.9,0.9),col=\"gray97\",lwd=8)\n");
    }
  else
    {
      fprintf(file,"lines(c(0,0.5),c(1.3,1.3),col=\"%s\",lwd=8)\n",COLOR1);
      fprintf(file,"lines(c(0,0.5),c(1.1,1.1),col=\"%s\",lwd=8)\n",COLOR2);
      fprintf(file,"lines(c(0,0.5),c(0.9,0.9),col=\"%s\",lwd=8)\n",COLOR3);
    }
  
  fprintf(file,"text(x=2.5,y=1.5,\"REGIONS\",cex=2,font=2)\n");
  fprintf(file,"text(x=2.65,y=1.3,\"PROMOTER\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[1],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=1.2,string,cex=1.5)\n");
  fprintf(file,"text(x=2.75,y=1.1,\"GENIC\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[2],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=1.0,string,cex=1.5)\n");
  fprintf(file,"text(x=2.75,y=0.9,\"INTERGENIC\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[3],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=0.8,string,cex=1.5)\n");
  fprintf(file,"rect(-0.15,0.7,5,1.6)\n");

  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(6,9,\"Figure legend. Distribution of ChIPseq reads across different regions of the genome.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.75,8.5,\"This plot is generated by counting the number of peaks fitting on each class of region.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.65,8,\"PROMOTER region is the region plus/minus %d bp upstream of the TSS and the TSS.\",font=1,cex=0.75)\n",SIMPLIFIED);
  fprintf(file,"text(5.6,7.5,\"GENIC region is the rest of intragenic region within the transcripts not overlapping with PROMOTER.\",font=1,cex=0.75)\n");
  fprintf(file,"text(4.225,7,\"INTERGENIC is the rest of the genome. TSS is the Transcription Start Site.\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,5.5,\"ChIPseq experiment:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,5,\"%s\",font=3,cex=0.75)\n",TrackName);
  fprintf(file,"text(6,4.5,\"Number of peaks:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,4,\"%ld\",font=3,cex=0.75)\n",nPeaks);
  fprintf(file,"text(6,3.5,\"RefSeq transcripts:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,3,\"%ld\",font=3,cex=0.75)\n",m->nTotalTranscripts);
  fprintf(file,"text(6,2.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,2,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"text(6,0,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  fprintf(file,"dev.off()\n\n");

  fclose(file);
}

/* Pie charts (detailed): peak genome distribution */
void CreateRscriptChartDetailed(char* RscriptFileName,
				char* PlotFileName,
				int nUtr5s,
				int nUtr3s,
				int nCDSs,
				int nDistals,
				int nProximals,
				int nIntrons,
				int nIntergenics,
				char* TrackName,
				long nPeaks)
{
  
  /* Output file */
  FILE* file;

  /* Total peaks */
  int nTotalPeaks;

  /* Messages */
  char mess[MAXSTRING];
    
  
  /* 0. Prepare one file to write */
  if ((file=fopen(RscriptFileName, "w"))==NULL)
    {
      sprintf(mess,"The R script commands File %s file can not be opened to write",
	      RscriptFileName);
      printError(mess);
    }

  /* Calculating the total number of peaks */
  nTotalPeaks =  nDistals + nProximals + nUtr5s + nUtr3s + nCDSs + nIntrons + nIntergenics;

  /* Pie chart representation */
  fprintf(file,"pdf(\"%s\")\n",PlotFileName); 
  fprintf(file,"slices<-c(%d,%d,%d,%d,%d,%d,%d)\n",nDistals,nProximals,nUtr5s,nUtr3s,nCDSs,nIntrons,nIntergenics);
  fprintf(file,"percent<-round(slices/sum(slices)*100,digits=1)\n");

  fprintf(file,"### FRAMEWORK\n");
  fprintf(file,"mat <- matrix(c(1,2),1)\n");
  fprintf(file,"layout(mat,c(2,1),c(1,1))\n");

  fprintf(file,"### LEFT PANEL\n");
  fprintf(file,"par(mar=c(0.5,0.5,0.5,0.5),lwd = 4)\n");

  if (GRADIENT)
    {
      fprintf(file,"colfunc<-colorRampPalette(c(\"%s\",\"white\"))\n",COLOR1);
      fprintf(file,"colors<-colfunc(7)\n");
      fprintf(file,"pie(slices,labels=NA, col=colors,");
    }
  else
    {
      fprintf(file,"pie(slices,labels=NA, col=c(\"%s\",\"%s\",\"%s\",\"%s\",\"%s\",\"%s\",\"%s\"),",
	      COLOR1,COLOR2,COLOR3,COLOR4,COLOR5,COLOR6,COLOR7);
    }
  fprintf(file,"init.angle=90,clockwise=TRUE,radius=1)\n");

  fprintf(file,"text(x=0,y=1.25,\"%s (%d peaks)\",cex=1.75,font=2)\n",TrackName,nTotalPeaks);
  GetTime(mess,m);
  fprintf(file,"text(x=0,y=-1.25,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);
  
  fprintf(file,"### RIGHT PANEL\n");
  fprintf(file,"plot(0,0,pch=\" \",ylim=c(0,2),xlim=c(0,5),axes=FALSE)\n");
  fprintf(file,"par(mar=c(0.5,0.5,0.5,0.5),lwd = 2)\n");

  if (GRADIENT)
    {
      fprintf(file,"lines(c(0,0.5),c(1.5,1.5),col=colors[1],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(1.3,1.3),col=colors[2],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(1.1,1.1),col=colors[3],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(0.9,0.9),col=colors[4],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(0.7,0.7),col=colors[5],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(0.5,0.5),col=colors[6],lwd=8)\n");
      fprintf(file,"lines(c(0,0.5),c(0.3,0.3),col=\"gray97\",lwd=8)\n");
    }
  else
    {
      fprintf(file,"lines(c(0,0.5),c(1.5,1.5),col=\"%s\",lwd=8)\n",COLOR1);
      fprintf(file,"lines(c(0,0.5),c(1.3,1.3),col=\"%s\",lwd=8)\n",COLOR2);
      fprintf(file,"lines(c(0,0.5),c(1.1,1.1),col=\"%s\",lwd=8)\n",COLOR3);
      fprintf(file,"lines(c(0,0.5),c(0.9,0.9),col=\"%s\",lwd=8)\n",COLOR4);
      fprintf(file,"lines(c(0,0.5),c(0.7,0.7),col=\"%s\",lwd=8)\n",COLOR5);
      fprintf(file,"lines(c(0,0.5),c(0.5,0.5),col=\"%s\",lwd=8)\n",COLOR6);
      fprintf(file,"lines(c(0,0.5),c(0.3,0.3),col=\"%s\",lwd=8)\n",COLOR7);
    }
  
  fprintf(file,"text(x=2.5,y=1.7,\"REGIONS\",cex=2,font=2)\n");
  fprintf(file,"text(x=2.65,y=1.5,\"DISTAL\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[1],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=1.4,string,cex=1.5)\n");

  fprintf(file,"text(x=2.75,y=1.3,\"PROXIMAL\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[2],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=1.2,string,cex=1.5)\n");

  fprintf(file,"text(x=2.75,y=1.1,\"5\'UTR\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[3],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=1.0,string,cex=1.5)\n");

  fprintf(file,"text(x=2.75,y=0.9,\"3\'UTR\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[4],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=0.8,string,cex=1.5)\n");

  fprintf(file,"text(x=2.75,y=0.7,\"CDS\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[5],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=0.6,string,cex=1.5)\n");

  fprintf(file,"text(x=2.75,y=0.5,\"INTRONS\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[6],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=0.4,string,cex=1.5)\n");

  fprintf(file,"text(x=2.75,y=0.3,\"INTERGENIC\",cex=1.5)\n");
  fprintf(file,"string<-paste(\"(\",percent[7],\"%%)\",sep=\"\")\n");
  fprintf(file,"text(x=2.65,y=0.2,string,cex=1.5)\n");
  fprintf(file,"rect(-0.15,0.1,5,1.6)\n");

  fprintf(file,"mat <- matrix(1,1)\n");
  fprintf(file,"layout(mat,c(1,1),c(1,1))\n");
  fprintf(file,"plot(0:10, type = \"n\", xaxt=\"n\", yaxt=\"n\", bty=\"n\", xlab = \"\", ylab = \"\")\n");
  fprintf(file,"text(6,9,\"Figure legend. Distribution of ChIPseq reads across different regions of the genome.\",font=2,cex=0.75)\n");
  fprintf(file,"text(5.75,8.5,\"This plot is generated by counting the number of peaks fitting on each class of region.\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.925,8,\"DISTAL region is the region within 2.5 Kbp and 0.5 Kbp upstream of the TSS. PROXIMAL\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.95,7.5,\"region is the region within 0.5 Kbp and the TSS. UTR is UnTRanslated sequence. CDS is\",font=1,cex=0.75)\n");
  fprintf(file,"text(5.8,7,\"the protein CoDing Sequence. INTRONS are intronic regions. INTERGENIC is the rest\",font=1,cex=0.75)\n");
  fprintf(file,"text(4.225,6.5,\"of the genome. TSS is the Transcription Start Site.\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,5.5,\"ChIPseq experiment:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,5,\"%s\",font=3,cex=0.75)\n",TrackName);
  fprintf(file,"text(6,4.5,\"Number of peaks:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,4,\"%ld\",font=3,cex=0.75)\n",nPeaks);
  fprintf(file,"text(6,3.5,\"RefSeq transcripts:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,3,\"%ld\",font=3,cex=0.75)\n",m->nTotalTranscripts);
  fprintf(file,"text(6,2.5,\"Window factor:\",font=1,cex=0.75)\n");
  fprintf(file,"text(6,2,\"%d\",font=3,cex=0.75)\n",WINDOWRES);
  fprintf(file,"text(6,0,cex=0.5,font=2,\"SeqCode by E. Blanco %s\")\n",mess);

  fprintf(file,"dev.off()\n\n");

  fclose(file);
}
